src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs

```diff
@@ -0,0 +1,348 @@
+                {
+                    if (inheritsFromIgnoringConstruction(current, originalContainingType))
+                    {
+                        if (originalThroughTypeOpt == null || inheritsFromIgnoringConstruction(originalThroughTypeOpt, current))
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
+            }
+
+            bool isPrivateSymbolAccessible(ISymbol within0, INamedTypeSymbol containingType)
+            {
+                var withinType = within0 as INamedTypeSymbol;
+
+                // A private symbol is accessible if we're (optionally nested) inside the type that it
+                // was defined in.
+                return withinType != null && isNestedWithinOriginalContainingType(withinType, containingType.OriginalDefinition);
+            }
+
+            bool isNestedWithinOriginalContainingType(INamedTypeSymbol type, INamedTypeSymbol possiblyContainingOriginalType)
+            {
+                Debug.Assert(possiblyContainingOriginalType.IsDefinition);
+                Debug.Assert(type != null);
+
+                for (var current = type.OriginalDefinition; current != null; current = current.ContainingType)
+                {
+                    if (sameOriginalNamedType(current, possiblyContainingOriginalType))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            bool inheritsFromIgnoringConstruction(ITypeSymbol type, INamedTypeSymbol baseType)
+            {
+                Debug.Assert(type.IsDefinition);
+                Debug.Assert(baseType.IsDefinition);
+
+                for (var current = type; current != null; current = current.BaseType?.OriginalDefinition)
+                {
+                    if (sameOriginalNamedType(baseType, current as INamedTypeSymbol))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            bool sameOriginalNamedType(INamedTypeSymbol t1, INamedTypeSymbol t2)
+            {
+                Debug.Assert(t1 != null);
+                Debug.Assert(t1.IsDefinition);
+                Debug.Assert(t2?.IsDefinition != false);
+
+                // We expect a given named type definition to satisfy reference identity.
+                // But we relax this to permit a type to be represented by multiple symbols (e.g. separate compilations).
+                // Note that the same symbol is expected to have the identical name as itself, despite VB language rules.
+                return t1 == t2 || t2 != null && t1.Name == t2.Name && t1.Arity == t2.Arity && sameOriginalSymbol(t1.ContainingSymbol, t2.ContainingSymbol);
```

Comment:
>t1.Name == t2.Name && t1.Arity == t2.Arity [](start = 49, length = 42)

To match the CLR rules, it is possible this should be `t1.MetadataName == t2.MetadataName`.